Add USB endpoint tests

fkjagodzinski opened the pull request:
### Description

Extend `tests-usb_device-basic` with:
* `endpoint test abort`,
* `endpoint test data correctness`,
* `endpoint test data toggle reset`,
* `endpoint test halt`,
* `endpoint test parallel transfers`,
* `endpoint test parallel transfers ctrl`.

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] New target
    [x] Feature
    [ ] Breaking change

CC @c1728p9 @maciejbocianski @jamesbeyond

fkjagodzinski committed:
USB: EndpointResolver: Add a generic method to get a free endpoint

fkjagodzinski committed:
Tests: USB: Generic: Add basic endpoint tests

fkjagodzinski commented:
The comments with test description are missing. I'll add them here or in the consecutive PR.

fkjagodzinski commented:
~The `endpoint test abort` test case needs #6884 to work.~

fkjagodzinski committed:
Tests: USB: Generic: Update documentation

c1728p9 commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -734,6 +812,350 @@ def control_stress_test(dev, log):
+        while time.time() < end_ts and not ctrl_error.is_set():
+            loopback_ep_test(ep_out, ep_in, ep_out.wMaxPacketSize)
+    except usb.core.USBError as err:
+        if err.errno not in (32, 110):
```
On windows this error is ```None``` when the endpoint is stalled. You might need to remove this check if the error codes aren't portable.

fkjagodzinski replied:
Done.

c1728p9 commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -734,6 +812,350 @@ def control_stress_test(dev, log):
+        while time.time() < end_ts:
+            halt_ep_test(dev, bulk_out, bulk_in, bulk_out, log)
+            bulk_out.clear_halt()
+            intf.set_altsetting()  # Force the device to start reading data from all OUT endpoints again.
```
I'm not sure if setting the alternate again is the right way to handle this, but I would like to get your thoughts on this @fkjagodzinski. Some possible alternatives
- Require that if a transfer was started then the transfer is resumed when halt is cleared (USBPhy responsible for this)
- Have the class driver handle the clear halt request and start the transfer again

fkjagodzinski replied:
It was the simplest solution I was thinking about. I haven't thought about the test class callback for a clear request though. I'll update the code this way.

fkjagodzinski replied:
I added `VENDOR_TEST_READ_START` device request. It's used to explicitly start reading on given OUT endpoint.

c1728p9 commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -734,6 +812,350 @@ def control_stress_test(dev, log):
+
+    delayed_halt = Timer(delay, timer_handler)
+    delayed_halt.start()
+    end_ts = time.time() + 1.5 * delay
```
This delay wasn't long enough for my machine and caused intermittent failures. It may be good to set this to a really long delay (1 second?), as it will break from the loop early due to the stall anyway.

fkjagodzinski replied:
Done, updated code does not use `end_ts` at all.

c1728p9 commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -734,6 +812,350 @@ def control_stress_test(dev, log):
+            log('Testing aborting an in progress transfer for IN endpoints.')
+        for ep_in in (bulk_in, interrupt_in):
+            payload_size = (NUM_PACKETS_UNTIL_ABORT + NUM_PACKETS_AFTER_ABORT) * ep_in.wMaxPacketSize
+            payload_in = ep_in.read(payload_size)
```
On my setup, Windows 10 using libusb-win32, this raises a timeout exception so ```payload_in``` never gets returned:
```
USBError: [Errno None] libusb0-dll:err [_usb_reap_async] timeout error
```

Maybe instead of aborting and stopping the transfer, abort and start a new transfer with a buffer with different contents. Alternatively, this could be broken up into two reads, where the first one should complete fully, and the second one should timeout without any data.

fkjagodzinski replied:
Done. Now reading data in chunks of `wMaxPacketSize` each.

fkjagodzinski commented:
I added the missing docstrings.

c1728p9 commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -734,6 +812,410 @@ def control_stress_test(dev, log):
+
+        if verbose:
+            log('Testing OUT/IN data correctness for bulk endpoint pair.')
+        for payload_size in range(1, bulk_out.wMaxPacketSize + 1):
```
Zero is a valid size, so it might not hurt to test that value as well. If you were seeing errors with this earlier #6922 should fix the errors.

fkjagodzinski replied:
Updated; thanks for the reminder!

fkjagodzinski committed:
Tests: USB: Update endpoint tests

Fix the host script issues present on Windows machines.
Add 0 B payload size to bulk endpoints test.
Update halt and abort tests according to PR comments.
Add an explicit request to start reading on OUT endpoints.

cmonr commented:
Umm... Is this PR still ongoing?

https://github.com/ARMmbed/mbed-os/pull/6884 is now closed...

fkjagodzinski commented:
@cmonr, Yes, I'm updating the code according to @c1728p9 comments.

fkjagodzinski commented:
Update:
* rebased to current `feature-hal-spec-usb-device`, 675b24eea,
* fixed host script issues present on Windows machines,
* added `0 B` payload size to bulk endpoints test,
* updated halt and abort tests according to PR comments,
* added an explicit request to start reading on OUT endpoints.

cmonr commented:
@c1728p9 Mind re-reviewing?

fkjagodzinski committed:
Tests: USB: Add data toggle test

c1728p9 commented:
This needs to be rebased but after that I will re-review. cc @fkjagodzinski

fkjagodzinski commented:
* rebased to current `feature-hal-spec-usb-device`, c753780,
* added `endpoint test data toggle reset`.

cmonr commented:
@c1728p9 Ping?

fkjagodzinski commented:
I consider this PR to be ready. The isochronous transfers are still missing -- I'll add them in a separeate PR.

c1728p9 commented on TESTS/usb_device/basic/USBEndpointTester.cpp:
```diff
@@ -0,0 +1,794 @@
+        _endpoint_buffs[i] = (uint8_t *) calloc(epc->max_packet, sizeof(uint8_t));
+        MBED_ASSERT(_endpoint_buffs[i] != NULL);
+    }
+    MBED_ASSERT(resolver.valid());
```
This assert is failing on the NUCLEO_F207ZG. It only has 1.25K of endpoint ram so I don't think it can handle this many at once at max size.

fkjagodzinski replied:
Decreased `TEST_SIZE_EP_ISO_MAX` to `256 B` as suggested.

cmonr commented:
/morph build

fkjagodzinski commented:
Build failures are a consequence of #7267 being merged to USB feature branch. I'll push an update shortly.

c1728p9 commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -837,6 +930,544 @@ def control_stress_test(dev, log):
+            raise_unconditionally(lineno(), 'USBError')
+
+
+def ep_test_abort(dev, log, verbose=False):
```
As part of this or another test, code should validate that after a transfer has been aborted the buffer isn't used.

fkjagodzinski replied:
I updated `ep_test_abort()` as suggested.

c1728p9 committed:
Update test to restart reads after unstall

After an endpoint is unstalled any pending transfers are terminated.
This patch re-starts any reads that were ongoing.

fkjagodzinski committed:
Tests: USB: Update ep callbacks

Endpoint callbacks no longer have endpoint as a param. This update was
introduced in #7267.

c1728p9 commented on TESTS/usb_device/basic/USBEndpointTester.cpp:
```diff
@@ -0,0 +1,817 @@
+    { false, USB_EP_TYPE_INT, TEST_SIZE_EP_INT_MAX, static_cast<ep_cb_t>(&USBEndpointTester::_cb_int_out) },
+    { true, USB_EP_TYPE_INT, TEST_SIZE_EP_INT_MAX, static_cast<ep_cb_t>(&USBEndpointTester::_cb_int_in) },
+    { false, USB_EP_TYPE_ISO, TEST_SIZE_EP_ISO_MAX, static_cast<ep_cb_t>(&USBEndpointTester::_cb_iso_out) },
+    { true, USB_EP_TYPE_ISO, TEST_SIZE_EP_ISO_MAX, static_cast<ep_cb_t>(&USBEndpointTester::_cb_iso_in) },
```
Changing the size to 256 bytes here allows the `NUCLEO_F207ZG` to run tests. Since ISO isn't being tested here anyway it might be a good quick fix.

c1728p9 commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -837,6 +930,544 @@ def control_stress_test(dev, log):
+                raise RuntimeError('Invalid endpoint status after halt operation')
+        except Exception as err:
+            ctrl_error.set()
+            log('Endpoint {:#04x} halt failed ({}).'.format(ep_to_halt.bEndpointAddress, err))
```
I'm seeing this error intermittently in the test logs, but the tests are still passing:
```
Endpoint 0x01 halt failed (Invalid endpoint status after halt operation).
```

It appears that sometimes the next test starts running and unhalts the endpoint before `timer_handler` calls get_status. If I wait for the timer to finish before returning this problem does not occur.

fkjagodzinski replied:
Good catch! I haven't spotted that. It appears I was naive enough to assume Python would kill a `Timer` thread created locally, when returning from a function.

fkjagodzinski replied:
Fixed https://github.com/ARMmbed/mbed-os/blob/1de5b2c3c6191fbb58ee40a574c2da62d88c5da0/TESTS/host_tests/pyusb_basic.py#L1050

fkjagodzinski committed:
Tests: USB: Decrease the max data size for iso ep

Although the USB spec sets the upper limit on FS isochronous endpoint
payloads to 1023 B, this value is hard to test in practice. Moreover,
not all the targets Mbed OS supports (like NUCLEO_F207ZG) are able to
handle all the endpoints set to max.

fkjagodzinski committed:
Tests: USB: Update endpoint halt test

Wait for a locally created Timer thread to finish before returning.

fkjagodzinski committed:
Tests: USB: Check ep buffer is released by abort

Validate that endpoint buffer is not used after a transfer has been
aborted.

fkjagodzinski commented:
* rebased to current `feature-hal-spec-usb-device`, 9165547
* updated the test code according to @c1728p9 feedback

0xc0170 commented:
/morph build

fkjagodzinski commented:
@cmonr @0xc0170 I can see that quite a few random test cases timed out. This PR only extends `tests-usb_device-basic`, so shouldn't have such a huge impact on other test suites. Can we retrigger `ci-morph-test`?

0xc0170 commented:
/morph test

0xc0170 commented:
Test results same. We had similar failures 2 weeks ago (no space left on device error), @cmonr please review

cmonr commented:
@studavekar FYI.

Devices power cycled. Restarting.
/morph test

0xc0170 reviewed (approved):
Note, please run astyle for this feature branch once the work is done  (might need to tweak astyleignore file))
